### PR TITLE
34 add data extraction result to space

### PR DESCRIPTION
**Pull request by giordano-lucas**
#34 #37

### CODE REVIEW

File: notte/env.py [hunk 1]
```diff
@@ -71,46 +71,68 @@ def previous_actions(self) -> list[Action] | None:
...
-        self._trajectory[-1] = obs  # update the last observation with the new space
-        return obs
+    async def _obslisting(self) -> Observation:
+        space = await self._context_to_action_space_pipe.forward_async(self.context, self.previous_actions)
+        async with self._obs_lock:
+            self.obs.space = space
+            return self.obs
+
+    async def _obsscraping(self) -> Observation:
+        # TODO: move this to `scrape` if this is not used more than once
```
Comment (giordano-lucas) [inline notte/env.py:102]: encore a faire ça

**Comment by giordano-lucas** [comment]
This PR also solves #37

File: notte/common/parser.py [hunk 2]
```diff
@@ -62,69 +79,128 @@ def search_pattern(text: str, tag: str) -> str | None:
...
         return EnvStepParams(action_id=action_id, params=params)
 
     @override
     def rules(self) -> str:
         return self.INSTRUCTIONS
 
-    @override
-    def textify(self, obs: Observation) -> str:
-        if obs.space is None:
+    def final_rules(self) -> str:
```
Comment (andreakiro) [inline notte/common/parser.py:134]: We have a variable `INSTRUCTIONS` that you can rename `PRE_INSTRUCTIONS` and move this function to another variable called `POST_INSTRUCTIONS` it makes more sense to me

File: notte/common/parser.py [hunk 3]
```diff
@@ -62,69 +79,128 @@ def search_pattern(text: str, tag: str) -> str | None:
...
                 return "rules"
 
+    @override
+    def is_done(self, text: str) -> bool:
+        return (
+            f"<{BaseNotteParser.done_tag}/>" in text or self.search_pattern(text, BaseNotteParser.done_tag) is not None
+        )
+
+    @override
+    def get_done_answer(self, text: str) -> str | None:
```
Comment (andreakiro) [inline notte/common/parser.py:103]: Why do you have a `is_done` and `get_done_answer` function that are essentially the same?

File: notte/env.py [hunk 4]
```diff
@@ -71,46 +71,68 @@ def previous_actions(self) -> list[Action] | None:
         if len(self._trajectory) <= 1:
             return None
         previous_obs: Observation = self._trajectory[-2]
-        if isinstance(previous_obs, PreObservation):
+        if not previous_obs.has_space():
             return None  # we don't have a space for pre-observations
-        if self.context.snapshot.clean_url != previous_obs.clean_url:
+        # if self.context.snapshot.clean_url != previous_obs.clean_url:
```
Comment (andreakiro) [inline notte/env.py:76]: Remove comment

File: notte/env.py [hunk 5]
```diff
@@ -71,46 +71,68 @@ def previous_actions(self) -> list[Action] | None:
...
         return preobs
 
-    def _obslisting(self, preobs: PreObservation) -> Observation:
-        space = self._context_to_action_space_pipe.forward(self.context, self.previous_actions)
-        obs = Observation(_url=preobs.url, _screenshot=preobs.screenshot, _space=space)
-        self._trajectory[-1] = obs  # update the last observation with the new space
-        return obs
+    async def _obslisting(self) -> Observation:
+        space = await self._context_to_action_space_pipe.forward_async(self.context, self.previous_actions)
+        async with self._obs_lock:
```
Comment (andreakiro) [inline notte/env.py:97]: This is ugly as hell. This function should take an observation as an input and tag it `obslisting(obs) -> Observation` and avoid at all cost to use an observation that is a mutable object from the class? I really see no point of doing this.

File: notte/env.py [hunk 6]
```diff
@@ -71,46 +71,68 @@ def previous_actions(self) -> list[Action] | None:
...
+            self.obs.space = space
+            return self.obs
+
+    async def _obsscraping(self) -> Observation:
+        # TODO: move this to `scrape` if this is not used more than once
+        data = await self._data_scraping_pipe.forward_async(self.context)
+        async with self._obs_lock:
```
Comment (andreakiro) [inline notte/env.py:104]: Same comment.

File: notte/env.py [hunk 7]
```diff
@@ -71,46 +71,68 @@ def previous_actions(self) -> list[Action] | None:
...
+            self.obs.data = data
+            return self.obs
+
+    async def _obsfull(self) -> Observation:
+        _ = await asyncio.gather(
```
Comment (andreakiro) [inline notte/env.py:109]: Hmmm OK I understand. But I am really NOT a fan of this behavior. Isn't there something thread-safer? And less far-fetched

File: notte/env.py [hunk 8]
```diff
@@ -71,46 +71,68 @@ def previous_actions(self) -> list[Action] | None:
...
+            self._context_to_action_space_pipe.forward_async(self.context, self.previous_actions),
+            self._data_scraping_pipe.forward_async(self.context),
+        )
+        return self.obs
 
     @timeit("goto")
-    async def goto(self, url: str) -> PreObservation:
+    async def goto(self, url: str | None = None) -> Observation:
```
Comment (andreakiro) [inline notte/env.py:116]: Why accepting None URLs? This is horrendous

File: notte/env.py [hunk 9]
```diff
@@ -71,46 +71,68 @@ def previous_actions(self) -> list[Action] | None:
...
+        preobs = Observation(url=snapshot.url, screenshot=snapshot.screenshot)
         self._trajectory.append(preobs)
         return preobs
 
-    def _obslisting(self, preobs: PreObservation) -> Observation:
-        space = self._context_to_action_space_pipe.forward(self.context, self.previous_actions)
-        obs = Observation(_url=preobs.url, _screenshot=preobs.screenshot, _space=space)
-        self._trajectory[-1] = obs  # update the last observation with the new space
-        return obs
+    async def _obslisting(self) -> Observation:
```
Comment (andreakiro) [inline notte/env.py:95]: This function is NOT async

File: notte/env.py [hunk 10]
```diff
@@ -71,46 +71,68 @@ def previous_actions(self) -> list[Action] | None:
...
+        space = await self._context_to_action_space_pipe.forward_async(self.context, self.previous_actions)
```
Comment (andreakiro) [inline notte/env.py:96]: Same here!
```suggestion
        space = self._context_to_action_space_pipe.forward_async(self.context, self.previous_actions)
```

File: notte/env.py [hunk 11]
```diff
@@ -71,46 +71,68 @@ def previous_actions(self) -> list[Action] | None:
...
 
     @timeit("observe")
-    async def observe(self, url: str) -> Observation:
-        preobs = await self.goto(url)
+    async def observe(self, url: str | None = None) -> Observation:
+        _ = await self.goto(url)
         logger.debug(f"ℹ️ previous actions IDs: {[a.id for a in self.previous_actions or []]}")
         logger.debug(f"ℹ️ context inodes IDs: {[node.id for node in self.context.interaction_nodes()]}")
-        return self._obslisting(preobs)
+        return await self._obslisting()
```
Comment (andreakiro) [inline notte/env.py:125]: Not async

File: notte/env.py [hunk 12]
```diff
@@ -119,39 +141,63 @@ async def execute(
         logger.info(f"🌌 action {action_id} executed in browser")
         return self._preobserve(snapshot)
 
+    @timeit("execute_special")
+    async def _execute_special(
```
Comment (andreakiro) [inline notte/env.py:145]: In the same way that we have an `ExecutionPipe` that allows to hide the boilerplate in this `env.py` file— i would do the same for the special actions

File: notte/llms/engine.py [hunk 13]
```diff
@@ -37,22 +37,32 @@ class StructuredContent:
...
+    def extract(self, text: str, fail_if_final_tag: bool = True, fail_if_inner_tag: bool = True) -> str:
         """Extract content from text based on defined tags"""
         content = text
 
         if self.outer_tag:
             pattern = f"<{self.outer_tag}>(.*?)</{self.outer_tag}>"
             match = re.search(pattern, content, re.DOTALL)
             if not match:
-                raise ValueError(f"No content found within <{self.outer_tag}> tags")
+                splits = text.split(f"<{self.outer_tag}>")
```
Comment (andreakiro) [inline notte/llms/engine.py:48]: What is this code?

File: notte/llms/engine.py [hunk 14]
```diff
@@ -37,22 +37,32 @@ class StructuredContent:
...
+                if fail_if_final_tag or len(splits) == 1:
```
Comment (andreakiro) [inline notte/llms/engine.py:49]: Same question here?

File: notte/llms/prompts/data-extraction/user.md [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+not not all websites have a navigation section, a footer section or a search results section.
+
+<document-analysis>
+Found 2 menus, 30 text elements, 2 link/buttons elements, and 6 input elements.
+Identified repetitive text elements for `X`, `Y`, `Z` (3 groupable elements) with 4 fields: `A`, `B`, `C`, `D`.
+Grouped text elements into 8 main categories based on ...
+[Additional analysis...]
+</document-analysis>
+<data-extraction>
+```markdown
```
Comment (andreakiro) [inline notte/llms/prompts/data-extraction/user.md:44]: Why do you always put these ```markdown``` tags?? They are useless. We already have the outer tag

File: notte/pipe/main.py [hunk 16]
```diff
@@ -46,3 +46,12 @@ def forward(
 
         actions = ActionFilteringPipe.forward(context, actions)
         return ActionSpace(_actions=actions)
+
+    async def forward_async(
```
Comment (andreakiro) [inline notte/pipe/main.py:50]: ???

**Review by andreakiro — CHANGES_REQUESTED** [review]

File: README.md [hunk 17]
```diff
@@ -54,43 +54,63 @@ os.environ['CEREBRAS_API_KEY'] = "your-api-key"
...
+...
 
-# Destination Exploration
-* L1: Shows flights from London to Tokyo
-* L2: Shows flights from New York to Rome
 [... More actions/categories ...]
 ```
 
-## As a backend for LLM web agents
+You can also scrape data from the page using the `scrape` function;
```
[the remark this answers is not in the document]
Comment (giordano-lucas) [reply README.md:87]: I disagree, it's an key functionality and I think we need to let users know they can do that. Anyone that wants to build an agent needs that functionality. Even if that's not the final version I think it's better to include that in the readme.

File: .vscode/settings.json [hunk 18]
```diff
@@ -26,5 +26,6 @@
   "[jsonc]": {
     "editor.defaultFormatter": "esbenp.prettier-vscode",
     "editor.formatOnSave": true
-  }
+  },
+  "python.languageServer": "None"
```
[the remark this answers is not in the document]
Comment (giordano-lucas) [reply .vscode/settings.json:30]: I think you should have the same locally. There is alway a pop up on cursor saying that we need that with pylance/pybasedright

File: notte/actions/space.py [hunk 19]
```diff
@@ -6,27 +6,43 @@
...
+            case ("all", "all"):
+                actions = self._actions
+            case ("all", _):
+                actions = [action for action in self._actions if action.role == role]
+            case (_, "all"):
+                actions = [action for action in self._actions if action.status == status]
+            case (_, _):
+                actions = [action for action in self._actions if action.status == status and action.role == role]
+
+        return actions + SpecialAction.list()  # type: ignore
```
[the remark this answers is not in the document]
Comment (giordano-lucas) [reply notte/actions/space.py]: It's outdated. And we do hit the return

File: notte/llms/engine.py (hunk 14 above)
Comment (giordano-lucas) [reply notte/llms/engine.py:49]: cf previous answer

**Review by andreakiro — COMMENTED** [review]
LGTM

**Pull request merged**